src/service/viewport/viewport-impl.js

```diff
@@ -999,25 +1013,82 @@ export class Viewport {
     }
   }
 
+  /**
+   * Specifies a callback for measuring an element's fixed offset when the
+   * viewport's padding top changes.
+   *
+   * `measure` runs in a measure phase and takes:
+   *  - afterAnimation
+   *    a promise that resolves after a running translation is finished.
+   *  - lastPaddingTop in px.
+   *  - paddingTop in px.
+   *
+   * And returns a vertical offset at start of animation. End will always be 0.
+   *
+   * This tears down the element from the fixed layer (but does NOT remove from
+   * the transfer layer), so its offset needs to be managed independently
+   * afterwards.
+   *
+   * @param {!Element} element
+   * @param {!FixedElementMeasureFnDef} measure
+   */
+  setFixedElementMeasurer(element, measure) {
+    this.removeFromFixedLayer(
+        element, /* onlyTearDown */ true, /* keepOffset */ true);
+    this.fixedMeasurers_.push({element, measure});
+  }
+
   /**
    * @param {number} duration
    * @param {string} curve
    * @param {boolean} transient
    * @return {!Promise}
    * @private
    */
-  animateFixedElements_(duration, curve, transient) {
-    this.fixedLayer_.updatePaddingTop(this.paddingTop_, transient);
-    if (duration <= 0) {
-      return Promise.resolve();
-    }
-    // Add transit effect on position fixed element
-    const tr = numeric(this.lastPaddingTop_ - this.paddingTop_, 0);
-    return Animation.animate(this.ampdoc.getRootNode(), time => {
-      const p = tr(time);
-      this.fixedLayer_.transformMutate(`translateY(${p}px)`);
-    }, duration, curve).thenAlways(() => {
-      this.fixedLayer_.transformMutate(null);
+  updateFixedElementsOffset_(duration, curve, transient) {
+    const {
+      paddingTop_: paddingTop,
+      lastPaddingTop_: lastPaddingTop,
+    } = this;
+
+    this.fixedLayer_.updatePaddingTop(paddingTop, transient);
+
+    let doneDeferred;
+
+    return this.vsync_.measurePromise(() => {
+      return this.fixedMeasurers_.map(def => {
+        doneDeferred = doneDeferred || new Deferred();
+        return def.measure(doneDeferred.promise, lastPaddingTop, paddingTop);
+      });
+    }).then(animOffsets => {
+      if (duration <= 0) {
+        return;
+      }
+
+      const interpolations =
+          animOffsets.map(animOffset => numeric(animOffset, 0));
+
+      const defaultAnimOffset = lastPaddingTop - paddingTop;
+      const defaultInterpolation = numeric(defaultAnimOffset, 0);
+
+      return Animation.animate(this.ampdoc.getRootNode(), time => {
+        this.fixedLayer_.transformMutate(
```

Comment:
Does this even need to be a part of fixed-layer?

Reply:
It was a part of fixed-layer before, because `fixed-layer` has knowledge of `tops`. If you're asking about a previous engineering decision, I'm not sure I can argue for it. I'm merely following the same API that was there before.

Reply:
I think we can rip this out of `FixedLayer`, which would be a code size win _and_ a code logic win.

Reply:
Hmm, `transformMutate` still has knowledge of `transform`s to apply the `translate` on top. It also knows which elements should be offset. It also serves as the single-source for elements that are mutated. If we don't track this in `FixedLayer`, we'd have to track it somewhere else and duplicate some logic. Maybe you could explain your desired architecture more thoroughly? Because in my head it sounds a little redundant.

Reply:
Going to move this discussion to https://github.com/ampproject/amphtml/pull/21279.

Reply:
👍